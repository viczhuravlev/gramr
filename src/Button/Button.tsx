import styled from '@emotion/styled';
import React, { StatelessComponent } from 'react';

interface IButton {
  text: string;
}

const SButton = styled.button`
  color: red;
`;

const Button: StatelessComponent<IButton> = ({ text, children }) => (
  <SButton>{text || children}</SButton>
);

export default Button;
