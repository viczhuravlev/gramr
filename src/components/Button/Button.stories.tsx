import React from 'react';
import { storiesOf } from '@storybook/react';

import { colors } from '../../styles/theme';
import { getColors, capitalizeFirstLetter } from '../../utils/helpers';

import { Button } from './index';

storiesOf('Button', module)
  .add('Default', () => <Button>Default</Button>)
  .add('Disabled', () => <Button isDisabled>isDisabled</Button>)
  .add('Colors', () => (
    <>
      {getColors().map((item) => (
        <Button color={colors[item]}>{capitalizeFirstLetter(item)}</Button>
      ))}
    </>
  ))
  .add('Size', () => (
    <>
      <Button size="s">Small</Button>
      <Button size="m">Medium</Button>
      <Button size="l">Large</Button>
    </>
  ))
  .add('Dark', () => (
    <>
      <Button theme="dark">Default</Button>
      <Button theme="dark" isDisabled>
        isDisabled
      </Button>
      {getColors().map((item) => (
        <Button color={colors[item]}>{capitalizeFirstLetter(item)}</Button>
      ))}
    </>
  ));
