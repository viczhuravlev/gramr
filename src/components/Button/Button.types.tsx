import { ReactNode } from 'react';

import { ThemeSize, ThemeColors, ThemeType } from '../../styles/theme.types';

export interface ButtonProps {
  size?: ThemeSize;
  theme?: ThemeType;
  color?: ThemeColors;
  onClick?: () => void;
  isDisabled?: boolean;
  className?: string;
  children?: ReactNode;
}
