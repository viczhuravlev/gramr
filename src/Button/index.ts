import Button from './Button';
import ButtonWithDescription from './ButtonWithDescription';

export { Button, ButtonWithDescription };
