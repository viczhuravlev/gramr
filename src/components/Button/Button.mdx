---
name: Button
---

import { Playground, PropsTable } from 'docz';

import Button from './Button';
import { colors } from '../../styles/theme';
import { capitalizeFirstLetter } from '../../utils/helpers';

# Button

<PropsTable of={Button} />

## Default

<Playground
  style={{ display: 'flex', justifyContent: 'space-between', flexWrap: 'wrap' }}
>
  <Button>Default</Button>
  <Button isDisabled>isDisabled</Button>
  {Object.keys(colors)
    .filter(item => item !== '__docgenInfo' && item !== 'displayName')
    .map(item => (
      <Button color={item}>{capitalizeFirstLetter(item)}</Button>
    ))}
</Playground>

## Dark

<Playground
  style={{
    display: 'flex',
    justifyContent: 'space-between',
    flexWrap: 'wrap',
    backgroundColor: '#1D2430'
  }}
>
  <Button theme="dark">Default</Button>
  <Button theme="dark" isDisabled>
    isDisabled
  </Button>
  {console.log(colors) ||
    Object.keys(colors)
      .filter(item => item !== '__docgenInfo' && item !== 'displayName')
      .map(item => <Button color={item}>{capitalizeFirstLetter(item)}</Button>)}
</Playground>

## Size

<Playground
  style={{ display: 'flex', justifyContent: 'space-between', flexWrap: 'wrap' }}
>
  <Button size="s">Small</Button>
  <Button size="m">Medium</Button>
  <Button size="l">Large</Button>
</Playground>
