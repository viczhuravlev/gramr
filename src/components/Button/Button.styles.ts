import { css } from '@emotion/core';
import styled from '@emotion/styled';

import { hexToRgb } from '../../utils/styles';
import { defineTheme } from '../../utils/theme';
import { ThemeLightKey } from '../../styles/theme.types';
import { theme, themeColors } from '../../styles/theme';

import * as T from './Button.types';

const sizeStyles = {
  s: css({
    height: '24px',
    padding: '0 7px',
    fontSize: '14px',
  }),
  m: css({
    height: '32px',
    padding: '0 15px',
    fontSize: '14px',
  }),
  l: css({
    height: '40px',
    padding: '0 20px',
    fontSize: '16px',
  }),
};

function getStylesTheme(
  props: T.ButtonProps,
  option: ThemeLightKey,
  optionDisabled: ThemeLightKey,
  colorsString?: string
) {
  const themeStrung = defineTheme(props.theme);

  if (props.isDisabled) {
    return theme[themeStrung][optionDisabled];
  }

  return props.color ? colorsString || props.color : theme[themeStrung][option];
}

export const Button = styled.button<T.ButtonProps>`
  ${(props) => sizeStyles[props.size || 'm'] || sizeStyles.m};
  border-radius: 4px;
  cursor: ${(props) => (props.isDisabled ? 'not-allowed' : 'pointer')};

  color: ${(props) =>
    getStylesTheme(props, 'font', 'fontDisabled', themeColors.white)};

  border: 1px solid
    ${(props) => getStylesTheme(props, 'border', 'borderDisabled')};

  background-color: ${(props) =>
    getStylesTheme(props, 'background', 'backgroundDisabled')};

  ${(props) =>
    props.isDisabled
      ? ''
      : `
        &:hover,
        &:focus,
        &:active {
          color: ${
            props.color
              ? themeColors.white
              : theme[defineTheme(props.theme)].fontHover
          };
          
          border-color: ${
            props.color
              ? props.color
              : theme[defineTheme(props.theme)].borderHover
          };
          
          background-color: ${props.color ? hexToRgb(props.color, 0.85) : ''};
         
          transition: color .15s ease-in-out, background-color .15s ease-in-out, border-color .15s ease-in-out;
        }
      `};
`;
