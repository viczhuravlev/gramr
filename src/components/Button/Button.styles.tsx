import styled from '@emotion/styled';

import { theme } from '../../styles/theme';
import { defineTheme } from '../../utils/theme';

import * as T from './Button.types';

export const Button = styled.button<T.ButtonProps>`
  height: 32px;
  padding: 0 15px;
  font-size: 14px;
  border-radius: 4px;
  cursor: ${props => (props.isDisabled ? 'not-allowed' : 'pointer')};

  // theme
  color: ${props => theme[defineTheme(props)][props.isDisabled ? 'textDisabled' : 'text']};
  border: 1px solid ${props => theme[defineTheme(props)].border}};
  background-color: ${props => theme[defineTheme(props)].background};
`;
