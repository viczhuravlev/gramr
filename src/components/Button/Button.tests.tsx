import React from 'react';
import { mount } from 'enzyme';

import Button from './Button';

describe('[component] Button', () => {
  it('[Snapshot] Button', () => {
    const component = mount(<Button color="orange">Test</Button>);

    expect(component).toMatchSnapshot();
  });
});
