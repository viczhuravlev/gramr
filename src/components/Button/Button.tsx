import React, { FC } from 'react';

import * as S from './Button.styles';
import * as T from './Button.types';

const Button: FC<T.ButtonProps> = ({
  size,
  theme,
  onClick,
  children,
  className,
  isDisabled,
  ...rest
}) => {
  if (!children) return null;

  return (
    <S.Button
      // size={size}
      theme={theme}
      disabled={isDisabled}
      isDisabled={isDisabled}
      className={className}
      onClick={isDisabled ? undefined : onClick}
      {...rest}
    >
      {children}
    </S.Button>
  );
};

Button.defaultProps = {
  size: 'm',
  theme: 'light',
  isDisabled: false
};

export default Button;
