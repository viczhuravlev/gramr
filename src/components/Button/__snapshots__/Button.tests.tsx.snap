// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[component] Button [Snapshot] Button 1`] = `
<Button
  color="orange"
  isDisabled={false}
  size="m"
  theme="light"
>
  <Styled(button)
    color="orange"
    disabled={false}
    isDisabled={false}
    size="m"
    theme="light"
  >
    <button
      className="css-107tu9w"
      color="orange"
      disabled={false}
      size="m"
    >
      Test
    </button>
  </Styled(button)>
</Button>
`;
